src/main/kotlin/pt/up/fe/ni/website/backend/service/RoleService.kt

```diff
@@ -0,0 +1,43 @@
+package pt.up.fe.ni.website.backend.service
+
+import org.springframework.stereotype.Service
+import pt.up.fe.ni.website.backend.model.Activity
+import pt.up.fe.ni.website.backend.model.PerActivityRole
+import pt.up.fe.ni.website.backend.model.Role
+import pt.up.fe.ni.website.backend.model.permissions.Permission
+import pt.up.fe.ni.website.backend.model.permissions.Permissions
+import pt.up.fe.ni.website.backend.repository.PerActivityRoleRepository
+import pt.up.fe.ni.website.backend.repository.RoleRepository
+
+@Service
+class RoleService(
+    private val roleRepository: RoleRepository,
+    private val perActivityRoleRepository: PerActivityRoleRepository
+) {
+
+    fun grantPermissionToRole(role: Role, permission: Permission) {
+        role.permissions.add(permission)
+        roleRepository.save(role)
+    }
+
+    fun revokePermissionFromRole(role: Role, permission: Permission) {
+        role.permissions.remove(permission)
+        roleRepository.save(role)
+    }
+
+    fun grantPermissionToRoleOnActivity(role: Role, activity: Activity, permission: Permission) {
+        val foundActivity = activity.associatedRoles
+            .find { it.activity == activity } ?: PerActivityRole(role, activity, Permissions())
```

Comment:
```suggestion
        val foundActivity = activity.associatedRoles
            .find { it.role == role } ?: PerActivityRole(role, activity, Permissions())
```

Shouldn't we filter based on the role, as we already know the activity is the requested one?

Reply:
I believe so. Perhaps this would be more readable:
```
val foundActivity = roles.perActivityRoles.find { it.activity == activity } ?: PerActivityRole(role, activity, Permissions())
```

Reply:
This also indicates the need for good testing when the controller is developed.